pkg/sql/rowexec/kv_scan.go

```diff
@@ -0,0 +1,176 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package rowexec
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+)
+
+// kvScan is used read raw KV values in a computation flow for purposes,
+// of mainly implementing crdb_internal.scan. This execinfra.RowSource
+// only returns a row format of key, value, and ts.
+type kvScan struct {
+	execinfra.ProcessorBase
+	spans           []roachpb.Span
+	rowIdx          int
+	bytesRead       int64
+	rowsRead        int64
+	scanResponses   []kvpb.ResponseUnion
+	currentResponse int
+	row             rowenc.EncDatumRow
+}
+
+// Start is part of the execinfra.RowSource interface.
+func (ks *kvScan) Start(ctx context.Context) {
+	ctx = ks.StartInternal(ctx, "index-scan")
+	// Set up the first batch.
+	ba := ks.EvalCtx.Txn.NewBatch()
+	for spanIdx := range ks.spans {
+		scan := kvpb.NewScan(ks.spans[spanIdx].Key,
+			ks.spans[spanIdx].EndKey,
+			false)
+		ba.AddRawRequest(scan)
+	}
+	ba.Header.MaxSpanRequestKeys = int64(rowinfra.ProductionKVBatchSize)
+	ba.Header.TargetBytes = int64(rowinfra.GetDefaultBatchBytesLimit(false))
+	err := ks.EvalCtx.Txn.Run(ctx, ba)
+	if err != nil {
+		ks.MoveToDraining(err)
+	}
+	ks.scanResponses = ba.RawResponse().Responses
+	// Set up the results for the rows.
+	ks.row = make(rowenc.EncDatumRow, 3)
+}
+
+// getNextScan will issue any additional scans to read additional data,
+// using any response spans.
+func (ks *kvScan) getNextScan(ctx context.Context) error {
+	// Generate a new batch with all of the resume spans
+	// from our first request.
+	ba := ks.EvalCtx.Txn.NewBatch()
+	for i := ks.currentResponse; i < len(ks.scanResponses); i++ {
+		scan := ks.scanResponses[i].GetScan()
+		nextSpan := scan.ResumeSpan
+		scanRequest := kvpb.NewScan(nextSpan.Key,
+			nextSpan.EndKey,
+			false /*forUpdate*/)
+		ba.AddRawRequest(scanRequest)
+	}
+	ba.Header.MaxSpanRequestKeys = int64(rowinfra.ProductionKVBatchSize)
+	err := ks.EvalCtx.Txn.Run(ctx, ba)
+	if err != nil {
+		return err
+	}
+	ks.scanResponses = ba.RawResponse().Responses
+	ks.currentResponse = 0
+	ks.rowIdx = 0
+	return nil
+}
+
+// Next is part of the execinfra.RowSource interface.
+func (ks *kvScan) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
+	if ks.State != execinfra.StateRunning {
+		return nil, ks.DrainHelper()
+	}
+	moreRows := true
+	var err error
```

Comment:
This isn't read outside of the for loop below so I'm not sure we need to declare it here.